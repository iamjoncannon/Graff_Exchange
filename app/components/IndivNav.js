import React from 'react';
import { isDesktop, isCell } from './utils'
import { Link } from 'react-router-dom'
import IndivSelector from './IndivSelector'
import TradeBox from './TradeBox'

export default class IndivNav2 extends React.Component {
 
  constructor(props) {
    super(props);

    this.state = {

        isModalShowing : false,
        whichModal : null // options || tradeBox
    }
  }
  
  closeModal = () => {

    this.setState({ isModalShowing : false })
  }

  render() {

    const {isModalShowing, whichModal} = this.state

    const { pathname } = this.props.location

    return (

    <div style={{height: "100vh"}}>
      
      <div className="individual-nav">
  
        { isCell() &&
       
          <i className="fas fa-bars fa-7x" 
             onClick={()=>this.setState({isModalShowing: true, whichModal: 'selector'})}
          />
        
        }

        <div>
  
          <span>AAPL</span> 
  
          
          { !isDesktop() && 
          
            <i className="fas fa-angle-down fa-7x" />
          }
  
        </div>
  
        <span     
          onClick={()=>this.setState({isModalShowing: true, whichModal: 'trade-box'})}
        >
          Trade
        </span>
      
      </div>

      {
        isModalShowing && 

        <div className="modal-container" onClick={()=>this.setState({isModalShowing: false})}>

          <div onClick={ (e)=> e.stopPropagation() } className={whichModal}>
            
              { whichModal === 'selector' ? 

                <IndivSelector 
                  pathname={pathname} 
                  exit={this.closeModal}
                />

                : 
                
                <TradeBox  
                  exit={this.closeModal}
                />
              }
          </div>
        </div>
      }
    </div>
    );
  }
}
