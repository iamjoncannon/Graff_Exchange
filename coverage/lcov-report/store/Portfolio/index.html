<!doctype html>
<html lang="en">
<head>
    <title>Code coverage report for store/Portfolio</title>
    <meta charset="utf-8" />
    <link rel="stylesheet" href="../../prettify.css" />
    <link rel="stylesheet" href="../../base.css" />
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <style type='text/css'>
        .coverage-summary .sorter {
            background-image: url(../../sort-arrow-sprite.png);
        }
    </style>
</head>
<body>
<div class='wrapper'>
  <div class='pad1'>
    <h1>
      <a href="../../index.html">All files</a> store/Portfolio
    </h1>
    <div class='clearfix'>
      <div class='fl pad1y space-right2'>
        <span class="strong">47.66% </span>
        <span class="quiet">Statements</span>
        <span class='fraction'>61/128</span>
      </div>
      <div class='fl pad1y space-right2'>
        <span class="strong">16% </span>
        <span class="quiet">Branches</span>
        <span class='fraction'>4/25</span>
      </div>
      <div class='fl pad1y space-right2'>
        <span class="strong">72.73% </span>
        <span class="quiet">Functions</span>
        <span class='fraction'>16/22</span>
      </div>
      <div class='fl pad1y space-right2'>
        <span class="strong">46.67% </span>
        <span class="quiet">Lines</span>
        <span class='fraction'>56/120</span>
      </div>
    </div>
    <p class="quiet">
      Press <em>n</em> or <em>j</em> to go to the next uncovered block, <em>b</em>, <em>p</em> or <em>k</em> for the previous block.
    </p>
  </div>
  <div class='status-line low'></div>
<div class="pad1">
<table class="coverage-summary">
<thead>
<tr>
   <th data-col="file" data-fmt="html" data-html="true" class="file">File</th>
   <th data-col="pic" data-type="number" data-fmt="html" data-html="true" class="pic"></th>
   <th data-col="statements" data-type="number" data-fmt="pct" class="pct">Statements</th>
   <th data-col="statements_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="branches" data-type="number" data-fmt="pct" class="pct">Branches</th>
   <th data-col="branches_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="functions" data-type="number" data-fmt="pct" class="pct">Functions</th>
   <th data-col="functions_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="lines" data-type="number" data-fmt="pct" class="pct">Lines</th>
   <th data-col="lines_raw" data-type="number" data-fmt="html" class="abs"></th>
</tr>
</thead>
<tbody><tr>
	<td class="file empty" data-value="action_constants_for_Portfolio.js"><a href="action_constants_for_Portfolio.js.html">action_constants_for_Portfolio.js</a></td>
	<td data-value="0" class="pic empty"><div class="chart"><div class="cover-fill" style="width: 0%;"></div><div class="cover-empty" style="width:100%;"></div></div></td>
	<td data-value="0" class="pct empty">0%</td>
	<td data-value="0" class="abs empty">0/0</td>
	<td data-value="0" class="pct empty">0%</td>
	<td data-value="0" class="abs empty">0/0</td>
	<td data-value="0" class="pct empty">0%</td>
	<td data-value="0" class="abs empty">0/0</td>
	<td data-value="0" class="pct empty">0%</td>
	<td data-value="0" class="abs empty">0/0</td>
	</tr>

<tr>
	<td class="file medium" data-value="actions_for_Portfolio.js"><a href="actions_for_Portfolio.js.html">actions_for_Portfolio.js</a></td>
	<td data-value="77.78" class="pic medium"><div class="chart"><div class="cover-fill" style="width: 77%;"></div><div class="cover-empty" style="width:23%;"></div></div></td>
	<td data-value="77.78" class="pct medium">77.78%</td>
	<td data-value="18" class="abs medium">14/18</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="0" class="abs high">0/0</td>
	<td data-value="55.56" class="pct medium">55.56%</td>
	<td data-value="9" class="abs medium">5/9</td>
	<td data-value="77.78" class="pct medium">77.78%</td>
	<td data-value="18" class="abs medium">14/18</td>
	</tr>

<tr>
	<td class="file low" data-value="reducer_for_Portfolio.js"><a href="reducer_for_Portfolio.js.html">reducer_for_Portfolio.js</a></td>
	<td data-value="6.98" class="pic low"><div class="chart"><div class="cover-fill" style="width: 6%;"></div><div class="cover-empty" style="width:94%;"></div></div></td>
	<td data-value="6.98" class="pct low">6.98%</td>
	<td data-value="43" class="abs low">3/43</td>
	<td data-value="10.53" class="pct low">10.53%</td>
	<td data-value="19" class="abs low">2/19</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="1" class="abs high">1/1</td>
	<td data-value="7.32" class="pct low">7.32%</td>
	<td data-value="41" class="abs low">3/41</td>
	</tr>

<tr>
	<td class="file medium" data-value="thunks_for_Portfolio.js"><a href="thunks_for_Portfolio.js.html">thunks_for_Portfolio.js</a></td>
	<td data-value="65.67" class="pic medium"><div class="chart"><div class="cover-fill" style="width: 65%;"></div><div class="cover-empty" style="width:35%;"></div></div></td>
	<td data-value="65.67" class="pct medium">65.67%</td>
	<td data-value="67" class="abs medium">44/67</td>
	<td data-value="33.33" class="pct low">33.33%</td>
	<td data-value="6" class="abs low">2/6</td>
	<td data-value="83.33" class="pct high">83.33%</td>
	<td data-value="12" class="abs high">10/12</td>
	<td data-value="63.93" class="pct medium">63.93%</td>
	<td data-value="61" class="abs medium">39/61</td>
	</tr>

</tbody>
</table>
</div><div class='push'></div><!-- for sticky footer -->
</div><!-- /wrapper -->
<div class='footer quiet pad2 space-top1 center small'>
  Code coverage
  generated by <a href="https://istanbul.js.org/" target="_blank">istanbul</a> at Thu Oct 03 2019 19:08:17 GMT-0400 (Eastern Daylight Time)
</div>
</div>
<script src="../../prettify.js"></script>
<script>
window.onload = function () {
        if (typeof prettyPrint === 'function') {
            prettyPrint();
        }
};
</script>
<script src="../../sorter.js"></script>
<script src="../../block-navigation.js"></script>
</body>
</html>
