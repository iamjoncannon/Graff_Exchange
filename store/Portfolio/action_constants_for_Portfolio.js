export default {
	HYDRATEPORTFOLIO : 'HYDRATEPORTFOLIO',
	MAKETRADE : 'MAKETRADE',
	GETONEPRICE : 'GETONEPRICE',
	GETOPENINGPRICE : 'GETOPENINGPRICE',

}