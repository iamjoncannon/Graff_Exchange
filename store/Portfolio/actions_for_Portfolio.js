import actions from "./action_constants_for_Portfolio"

const hydratePortfolio = (payload) => {
      	return {
      		type: actions.HYDRATEPORTFOLIO,
      		payload
      	}
      }

const makeTrade = (payload) => {
      	return {
      		type: actions.MAKETRADE,
      		payload
      	}
      }

const getOnePrice = (payload) => {
      	return {
      		type: actions.GETONEPRICE,
      		payload
      	}
      }

const getOpeningPrice = (payload) => {
      	return {
      		type: actions.GETOPENINGPRICE,
      		payload
      	}
      }

export default {
	hydratePortfolio,
	makeTrade,
	getOnePrice,
	getOpeningPrice,
}