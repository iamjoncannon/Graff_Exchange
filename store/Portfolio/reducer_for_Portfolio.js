import actions from "./action_constants_for_PORTFOLIO"

const initialState = {
  selectedPortfolioItem: null,
  portfolio: { },
  transactionHistory: { }
}

export default function Portfolio_reducer (state = initialState, action) {
  
  switch (action.type) {

    case actions.HYDRATEPORTFOLIO: {

      const { portfolio, transactionHistory } = action.payload
      
      let selectedPortfolioItem = !state.selectedPortfolioItem ? portfolio[Object.keys(portfolio)[0]] : selectedPortfolioItem ;

      return { portfolio, transactionHistory, selectedPortfolioItem }
    }
    
    case actions.MAKETRADE: {

      let { symbol, trade } = action.payload

      let updatedPortfolio = {...state.portfolio } 

      updatedPortfolio[symbol]["quantity"] = trade[1]

      return { ...state, portfolio: updatedPortfolio  }
    }
    
    case actions.GETONEPRICE: {

      return { ...state }
    }
    
    case actions.GETOPENINGPRICE: {

      return { ...state }
    }
    
    case actions.HANDLESOCKETMESSAGE: {

      const stock = action.payload

      let updatedPortfolio = {...state.portfolio } 

      updatedPortfolio[stock.symbol]["price"] = stock.price

      return { ...state, portfolio: updatedPortfolio }
    }
    
    case actions.HANDLENEWS: {

      const {symbol, data } = action.payload
      let updatedPortfolio = { ...state.portfolio}

      updatedPortfolio[symbol]["news"] = data.data;
  
      return { ...state, portfolio: updatedPortfolio }
    }
    
    case actions.HANDLEFINANCIALS: {

      const {symbol, financials } = action.payload

      let updatedPortfolio = { ...state.portfolio}

      updatedPortfolio[symbol]["financials"] = financials[0];
  
      return { ...state, portfolio: updatedPortfolio }
    }
    
    case actions.HANDLEHISTORICALPRICE: {

      const {symbol, historical } = action.payload
      
      let updatedPortfolio = { ...state.portfolio}

      updatedPortfolio[symbol]["historical"] = historical;
  
      return { ...state, portfolio: updatedPortfolio }
    }
    
    case actions.HANDLESYMBOLSELECT: {

      const nextSelect = state.portfolio[action.payload]

      return { ...state, selectedPortfolioItem : nextSelect }
    }
    
    default:
      return state
  }
}