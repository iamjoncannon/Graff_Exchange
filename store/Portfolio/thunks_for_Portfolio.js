import actions from "./actions_for_Portfolio"
import { urlPrefix } from '../../secrets'
import axios from 'axios'

function makeHeader(token){
  
  return { headers: {"Authorization": 'Bearer ' + token}}
}

// local storage has no TTL mechanism like Redis,
// so we have to roll our own 

function epoch(){

  // Unix epoch converted hours
  return ( (new Date).getTime() / 1000 ) / (60*60)
}

function setTTL(hours){

  return epoch() + hours
}

function hasTTLExpired(endTime){

  let current = epoch()

  return current > endTime
}

export const hydratePortfolioThunk = (token) => async dispatch => {
 
  let portfolio
  let transactionHistory

  try {

    Promise.all( // allows us to run api calls in parallel rather than serially
      [ 
        ( portfolio = await axios.post( urlPrefix + '/getportfolio', {}, makeHeader(token)) ),
        ( transactionHistory = await axios.post(urlPrefix + '/getallTransactions', {}, makeHeader(token)) )
      ]
    )
  }
  catch(error){

    console.log(error)
  }
  
  portfolio = JSON.parse(portfolio.data)
  
  // prevent nullish errors

  for(let stock in portfolio){
    
    portfolio[stock].data = {}
  }
  
  // update the store with what we have now, so it can
  // start updating the table- this will cause the view
  // to update gradually rather than waiting for all the data
  // to return from the api in the next calls

  dispatch(actions.hydratePortfolio( { portfolio : {...portfolio}, 
                                       transactionHistory: {...JSON.parse(transactionHistory.data)},                 
                                      //  financials : null,
                                      //  historical : null,
                                      //  news : null
                                      } 
                                     ))
  
  // we have the initial portfolio, now we need to add current data
  // from an external API using an endpoint from the Gopher API

  if(portfolio && transactionHistory.data){

    for(let stock in portfolio){
                                        // hoisted below
        portfolio[stock].data = await getOpeningPriceThunk(stock, token)

    }

    dispatch(actions.hydratePortfolio( { portfolio : {...portfolio}, 
                                         transactionHistory: {...JSON.parse(transactionHistory.data)},                 
                                        } ))
  }
  else{

    // registration user story
    // callback({}, {1:{Symbol: "", Quantity: "", Date: ""}})

  }
  
};


// data cached in local storage

export const hydrateSinglePortfolioPage = (token, selectedPortfolioItem) => async dispatch => {

  let { symbol } = selectedPortfolioItem

  // basic pattern of implementing local storage with Unix epoch TTL

  // data for the time series chart

  {

    let inLocalStorage = !!localStorage.getItem(`time-series-${symbol}`)  
    
    let expired = hasTTLExpired(localStorage.getItem(`time-series-${symbol}-TTL`)) 
    
    if( inLocalStorage && !expired ){
      
      dispatch(actions.handleFinancials({ symbol, financials: JSON.parse(localStorage.getItem(`time-series-${symbol}`))}))
    }
    else{

      let data 

      try{

        data = await axios.post( urlPrefix + `/time-series/${symbol}`, {}, makeHeader(token) ) 

      }
      catch(error){

        console.log(error)
      }
        
      if(data){
        
        let { historical } = data.data
        dispatch(actions.handleHistoricalPrice({ symbol, historical }))
        localStorage.setItem(`time-series-${symbol}`, JSON.stringify(historical))
        localStorage.setItem(`time-series-${symbol}-TTL`, setTTL(60*60*24) )
      }        
    }
  }
  
  // data for the news section
  
  {

    let inLocalStorage = !!localStorage.getItem(`news-${symbol}`)  // returns null if key doesn't exist
    
    let expired = hasTTLExpired(localStorage.getItem(`news-${symbol}-TTL`)) 
    
    if( inLocalStorage && !expired ){
      
      dispatch(actions.handleNews({ symbol, data: JSON.parse(localStorage.getItem(`news-${symbol}`))}))
    }
    else{
      
      // calling Gopher API
      let data 
      
      try{
        
        let response = await axios.post( urlPrefix + `/news/${symbol}`, {}, makeHeader(token) ) 

        data = response.data
      }
      catch(error){

        console.log(error)
      }
                
      if(data){
        
          dispatch(actions.handleNews({ symbol, data }))
          localStorage.setItem(`news-${symbol}`, JSON.stringify(data))
          localStorage.setItem(`news-${symbol}-TTL`, setTTL(.30) )
        }
      }
  }


  // data for the quarterly financials section
  {

    let inLocalStorage = !!localStorage.getItem(`financials-${symbol}`)  
    
    let expired = hasTTLExpired(localStorage.getItem(`financials-${symbol}-TTL`)) 
    
    if( inLocalStorage && !expired ){
      
      dispatch(actions.handleFinancials({ symbol, financials: JSON.parse(localStorage.getItem(`financials-${symbol}`))}))
    }
    else{

      let data 

      try{

        data = await axios.post( urlPrefix + `/financials/${symbol}`, {}, makeHeader(token) ) 

      }
      catch(error){

        console.log(error)
      }
        
      if(data){
        
        let { financials } = data.data
        dispatch(actions.handleFinancials({ symbol, financials }))
        localStorage.setItem(`financials-${symbol}`, JSON.stringify(financials))
        localStorage.setItem(`financials-${symbol}-TTL`, setTTL(60*60*24) )
      }        
    }
  }

  

  
  // const timeSeriesData = `https://financialmodelingprep.com/api/v3/historical-price-full/${symbol}?timeseries=365`

  // axios.get(timeSeriesData).then( ({ data }) => {

  //   const { historical } = data

  //   dispatch(actions.handleHistoricalPrice({ symbol, historical }))
  // })

}

export const makeTradeThunk = (symbol, Quantity, Type, Price, token) => async dispatch => {

  /*
    endpoint expects:
     Symbol: '',
     Quantity: '',
     Type: 'Buy',
     Price: ''
  */
 
    Quantity = Number(Quantity)

    let trade 

    try{

      trade = await axios.post( urlPrefix + '/maketransaction', {Symbol: symbol, Quantity, Type, Price}, makeHeader(token)) 
    }
    catch(err){

      console.log(err)
    }

    trade = trade.data

    dispatch(actions.makeTrade({symbol, trade})) 
};

async function getOpeningPriceThunk (symbol, token) {
  
  const url = urlPrefix + '/ohlc/' + symbol

  let returnData  

  try {

    let { data } = await axios.post(url, {}, makeHeader(token))

    returnData = data

  }
  catch(error){
    console.log(error)
  }

  // returned from redis cache

  if(typeof returnData === "string"){

    try{

      returnData = JSON.parse(returnData)
    }
    catch(error){
      console.log("error reading JSON: ", symbol, error)
    }

  }
  
  return returnData
}

export default {
	hydratePortfolioThunk,
	makeTradeThunk,
}
