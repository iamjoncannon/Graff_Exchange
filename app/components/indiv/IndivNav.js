import React from 'react';
import { connect } from "react-redux";
import { Link } from 'react-router-dom'
import { isDesktop, isCell, isTab } from '../utils'
import IndivSelector from './IndivSelector'
import TradeBox from './TradeBox'
import { hydrateSinglePortfolioPage } from "../../../store/Portfolio/thunks_for_Portfolio.js"
import actions  from '../../../store/Portfolio/actions_for_Portfolio.js'

// this component manages hydrating data
// for the individual stock section
// as well as the navigational logic of
// that section

class IndivNav extends React.Component {
 
  constructor(props) {
    super(props);

    this.state = {
        whichItem: null,
        isModalShowing : false,
        whichModal : null // options || tradeBox
    }
  }

  componentDidMount(){
        
      this.setState({
          whichItem : this.props.selectedPortfolioItem.symbol
      })

      this.updateData()
  }

  componentDidUpdate(){

      if(this.props.selectedPortfolioItem.symbol !== this.state.whichItem){
          
          this.setState({
              whichItem: this.props.selectedPortfolioItem.symbol
          })

          this.updateData()
      }
  }

  updateData(){

      const { selectedPortfolioItem, token } = this.props

      this.props.hydrateSinglePortfolioPage(token, selectedPortfolioItem, this.props.portfolio)

  }

  closeModal = () => {

    this.setState({ isModalShowing : false })
  }

  render() {

    const { isModalShowing, whichModal } = this.state
    const { pathname } = this.props.location
    const { selectedPortfolioItem, portfolio, handleSymbolSelect } = this.props

    return (

      <div>

      <div className="individual-nav">
  
        { isCell() ? 
       
          <i className="fas fa-bars fa-7x" 
             onClick={()=>this.setState({isModalShowing: true, whichModal: 'page-selector'})}
          />
          :
          <div style={{width: isDesktop()? "0vw" : "7vw"}}></div>
        
        }

        <div  
          onClick={()=>this.setState({isModalShowing: true, whichModal: 'symbol-selector'})}
        >
  
          <span>{selectedPortfolioItem && selectedPortfolioItem.symbol}</span> 
  
          
          { !isDesktop() && 
          
            <i className="fas fa-angle-down fa-7x" />
          }
  
        </div>
  
        <span     
          onClick={()=>this.setState({isModalShowing: true, whichModal: 'trade-box'})}
        >
          Trade
        </span>
      
      </div>
      {
        !isCell() &&
        
        <div className="selector">
        
          <IndivSelector 
            pathname={pathname} 
            exit={()=>{}}
          />

        </div>
      
      }

      {
        isModalShowing && 

        <div className="modal-container" onClick={()=>this.setState({isModalShowing: false})}>

          <div onClick={ (e)=> e.stopPropagation() } className={whichModal}>
            
              { whichModal === 'page-selector' &&

                <IndivSelector 
                  pathname={pathname} 
                  exit={this.closeModal}
                />
              }

              { whichModal === 'trade-box' && 
                
                <TradeBox  
                  exit={this.closeModal}
                />
              }

              {
                whichModal === 'symbol-selector' && portfolio &&

                
                
                  <div onClick={(e)=> { 
                      handleSymbolSelect(e.target.textContent)
                      this.setState({isModalShowing: false})
                    }} > 

                    {Object.entries(portfolio).map((item)=>{

                     

                        return(
                          <span>{item[0]}</span>
                        )
                    })}
                  </div>
                
              }

          </div>
        </div>
      }

    </div>
    );
  }
}

const mapStateToProps = ({ Portfolio_state, User_state }) => {
  return {
    token: User_state.token,
    selectedPortfolioItem: Portfolio_state.selectedPortfolioItem,
    portfolio: Portfolio_state.portfolio
  };
};

const mapDispatchToProps = dispatch => ({
  hydrateSinglePortfolioPage: (...args) => dispatch(hydrateSinglePortfolioPage(...args)),
  handleSymbolSelect: (symbol) => dispatch(actions.handleSymbolSelect(symbol)),
});

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(IndivNav);