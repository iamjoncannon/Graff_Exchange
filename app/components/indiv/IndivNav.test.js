import React from 'react';
import renderer from 'react-test-renderer';
import { Provider } from 'react-redux';
import IndivNav from './IndivNav'
import { mount, shallow } from 'enzyme'
import configureStore from 'redux-mock-store';
import thunk from "redux-thunk"
const mockStore = configureStore([thunk]);
import { hydrateSinglePortfolioPage } from "../../../store/Portfolio/thunks_for_Portfolio"
import nocked from "../../../test/nock.setup.js"
import { BrowserRouter as Router } from "react-router-dom";

/*

tests

unit test (display)- 

renders a span with the selected portfolio item and an 
angle down icon 
    - if you click, sets state to symbol selector modal

renders a span that says Trade 
    - if you click on Trade, sets state to Trade modal

if cell, renders an "fa-bars" icon, 
    - if you click, sets state to page selector modal 

if not cell, 
    renders a holder div with a specific length
    renders an Indiv Selector component

integration tests- 

if historical not defined, and viewport is cell phone,
calls hydrateSinglePortfolioPage with fake data
and dispatches to the store 

* if modal showing

renders container

    - if you click the container, it exits the modal

    if modal is page, 
        renders individual selector
        - if you click on indiv exit, it exits the modal

    if modal is trade, 
        renders trade box
        - if you click on trade exit, it exits the modal
        
    if modal is symbol, 
        renders div 
        - if clicked, handlesymbol selector callback and setstate
        to modal false 
        renders list of portfolio's symbols


*/

const fake_store_with_historical_data =  { 

    Portfolio_state: {
        portfolio: {
            FB: {
                symbol: "FB",
                historical: []
            },
            GOOGL: {

            },
            AAPL: {

            }
        },
        selectedPortfolioItem: "FB"
    }, 
    User_state: {

        balance: 0
    }
}

const fake_store_without_historical_data =  { 

    Portfolio_state: {
        portfolio: {
            FB: {
                symbol: "FB",
                historical: []
            }
        },
        selectedPortfolioItem: "FB"
    }, 
    User_state: {

        balance: 0
    }
}

describe("IndivNav", ()=>{

    let component 
    let connected_react_component
    let store 

    beforeEach(()=>{
        
        store = mockStore( fake_store_with_historical_data );

        component = renderer.create(
            <Provider store={store} >
                <Router >
                    <IndivNav location={{pathname: "/"}}/>
                </Router>
            </Provider>
        );

        // the actual component is contained within several wrappers
        connected_react_component = component.root.children[0].children[0].children[0].children[0]._fiber.stateNode
    })

    it('1) renders successfully with data from redux store', ()=>{

        expect(true).toBeTruthy()

        let tree = component.toJSON();
        
        expect(tree).toMatchSnapshot();
    })

    it("2) renders a span with the selected portfolio item and an angle down icon", ()=>{
        
        let className = "fas fa-angle-down fa-7x"
    
        let angle = component.root.find(node => node.props.className === className  )

        expect(angle).toBeTruthy()

        let sibling = angle.parent.props.children[0]

        expect(sibling.props.children).toEqual("FB")   
    })
    
    it("3) clicking the angle's parent div sets state to symbol selector modal",()=>{

        let className = "fas fa-angle-down fa-7x"
    
        let angle = component.root.find(node => node.props.className === className  )

        renderer.act(()=>{

            angle.parent.props.onClick()
        })

        expect(connected_react_component.state.isModalShowing).toEqual(true)
        expect(connected_react_component.state.whichModal).toEqual('symbol-selector')
    })

    it("4) renders a span containing 'Trade'", ()=>{
        
        let trade_span = component.root.find(node => node.props.children === "Trade")

        expect(trade_span).toBeTruthy()
        expect(trade_span.type).toEqual("span")
    })
    
    it("5) clicking trade sets state to Trade modal", ()=>{
        
        let trade_span = component.root.find(node => node.props.children === "Trade")

        renderer.act(()=>{

            trade_span.props.onClick()
        })

        expect(connected_react_component.state.isModalShowing).toEqual(true)
        expect(connected_react_component.state.whichModal).toEqual('trade-box')
    })
    
    it("6) if not cell viewport, renders Indiv Selector component", ()=> {
    
        let indiv_selector = component.root.find( node => node.type.name === "IndivSelector")

        expect(indiv_selector).toBeTruthy()
    })

})

describe("IndivNav", ()=>{

    let component 
    let connected_react_component
    let store 

    beforeEach(()=>{
        
        window.outerWidth = 600

        store = mockStore( fake_store_with_historical_data );

        component = renderer.create(
            <Provider store={store} >
                <Router >
                    <IndivNav location={{pathname: "/"}}/>
                </Router>
            </Provider>
        );

        // the actual component is contained within several wrappers
        connected_react_component = component.root.children[0].children[0].children[0].children[0]._fiber.stateNode
    })

    it(">> if cell viewport <<", ()=>{})

    it("7) renders bars icon", ()=>{

        const className = "fas fa-bars fa-7x" 

        const bars = component.root.find(node => node.props.className === className  )

        expect(bars).toBeTruthy()
    })

    it("8) if props.portfolio.historical is defined, does not call hydrateSinglePortfolioPage", ()=>{

        store.dispatch( hydrateSinglePortfolioPage("FB") )
            .then(()=> {
                
                expect(dispatched_actions.length).toEqual(1)
                    
                done()
            })             
    })
})

describe("IndivNav", ()=>{

    let component 
    let connected_react_component
    let store 

    beforeEach(()=>{
        
        window.outerWidth = 600

        store = mockStore( fake_store_without_historical_data );

        component = renderer.create(
            <Provider store={store} >
                <Router >
                    <IndivNav location={{pathname: "/"}}/>
                </Router>
            </Provider>
        );

        // the actual component is contained within several wrappers
        connected_react_component = component.root.children[0].children[0].children[0].children[0]._fiber.stateNode
    })

    it("9) if historical not defined, calls hydrateSinglePortfolioPage", ()=>{

        store.dispatch( hydrateSinglePortfolioPage("FB") )
            .then(()=> {
                
                expect(dispatched_actions.length).toEqual(2)
                expect(dispatched_actions[0].type).toEqual(dispatched_actions[1].type)
                    
                done()
            })   
    })
})

describe("IndivNav", ()=>{

    let component 
    let connected_react_component
    let store 

    beforeEach(()=>{
        
        window.outerWidth = 600

        store = mockStore( fake_store_with_historical_data );

        component = renderer.create(
            <Provider store={store} >
                <Router >
                    <IndivNav location={{pathname: "/"}}/>
                </Router>
            </Provider>
        );

        // the actual component is contained within several wrappers
        connected_react_component = component.root.children[0].children[0].children[0].children[0]._fiber.stateNode
    })

    it(">> if modal showing <<", ()=>{})

    it("10) if modal is page-selector, renders individual selector", ()=>{

        renderer.act(()=>{

            connected_react_component.setState({ isModalShowing: true, whichModal: "page-selector" }) 
         
        })

        let indiv_selector = component.root.find( node => node.type.name === "IndivSelector")
 
        expect(indiv_selector).toBeTruthy()
    })

    it("11) clicking individual selector exit prop closes the modal", ()=>{

        renderer.act(()=>{

            connected_react_component.setState({ isModalShowing: true, whichModal: "page-selector" }) 
        })
        
        let indiv_selector = component.root.find( node => node.type.name === "IndivSelector")
        
        renderer.act(()=>{

            indiv_selector.props.exit()
        })
 
        expect(connected_react_component.state.isModalShowing).toEqual(false)
    })

    it("12) if modal is trade-box, renders trade box", ()=>{

        renderer.act(()=>{

            connected_react_component.setState({ isModalShowing: true, whichModal: 'trade-box' }) 
        })
        
        let trade_box = component.root.find( node => node.type.name === "TradeBox")

        expect(trade_box).toBeTruthy()
        
    })

    it("13) clicking trade box exit prop closes the modal", ()=>{

        renderer.act(()=>{

            connected_react_component.setState({ isModalShowing: true, whichModal: 'trade-box' }) 
        })
        
        let trade_box = component.root.find( node => node.type.name === "TradeBox")
        
        renderer.act(()=>{

            trade_box.props.exit()
        })
 
        expect(connected_react_component.state.isModalShowing).toEqual(false)
    })

    it("14) if modal is symbol-selector, renders div with list of portfolio's symbols", ()=>{
        
        renderer.act(()=>{

            connected_react_component.setState({ isModalShowing: true, whichModal: 'symbol-selector' }) 
        })

        let elements = component.root.findAllByType( "span" ).map(el=>el.props.children)
                
        expect(elements.includes("AAPL")).toBeTruthy()
        expect(elements.includes("GOOGL")).toBeTruthy()
        expect(elements.includes("FB")).toBeTruthy()
    })

    it("15) clicking the div sets state to modal false", ()=>{

        renderer.act(()=>{

            connected_react_component.setState({ isModalShowing: true, whichModal: 'symbol-selector' }) 
        })

        let final_div = component.root.findAllByType("div")

        renderer.act(()=>{

            final_div[final_div.length-1].props.onClick({target: {textContent: null}})
        })
        
        expect(connected_react_component.state.isModalShowing).toBe(false)
    })
})
