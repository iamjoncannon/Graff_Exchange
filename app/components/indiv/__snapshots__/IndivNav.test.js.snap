// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IndivNav 1) renders successfully with data from redux store 1`] = `
<div>
  <div
    className="individual-nav"
  >
    <div
      style={
        Object {
          "width": "7vw",
        }
      }
    />
    <div
      onClick={[Function]}
    >
      <span>
        FB
      </span>
      <i
        className="fas fa-angle-down fa-7x"
      />
    </div>
    <span
      onClick={[Function]}
    >
      Trade
    </span>
  </div>
  <div
    className="page-selector"
  >
    <div
      onClick={[Function]}
    >
      <a
        href="/indiv/perf"
        onClick={[Function]}
      >
        <span>
          Performance
        </span>
      </a>
      <a
        href="/indiv/finan"
        onClick={[Function]}
      >
        <span>
          Financials
        </span>
      </a>
      <a
        href="/indiv/news"
        onClick={[Function]}
      >
        <span>
          News
        </span>
      </a>
      <a
        href="/indiv/trans"
        onClick={[Function]}
      >
        <span>
          Transactions
        </span>
      </a>
    </div>
  </div>
</div>
`;
